File: mobile/types/shop.ts
/**
 * Shop Type Definitions
 *
 * TypeScript interfaces and enums for the e-commerce shop system
 * with AI-powered product recommendations.
 */

// ===== ENUMS =====

export enum ProductCategory {
  SUPPLEMENTS = 'supplements',
  EQUIPMENT = 'equipment',
  APPAREL = 'apparel',
  ACCESSORIES = 'accessories',
  NUTRITION = 'nutrition',
  RECOVERY = 'recovery',
  TECH = 'tech',
  OTHER = 'other',
}

export enum OrderStatus {
  PENDING = 'pending',
  PROCESSING = 'processing',
  SHIPPED = 'shipped',
  DELIVERED = 'delivered',
  CANCELLED = 'cancelled',
  REFUNDED = 'refunded',
}

// ===== PRODUCT INTERFACES =====

export interface Product {
  id: number;
  name: string;
  slug: string;
  description?: string;
  short_description?: string;
  category: ProductCategory;
  brand?: string;
  price: number;
  compare_at_price?: number;
  sku?: string;
  stock_quantity: number;
  weight_kg?: number;
  features?: string[];
  specifications?: Record<string, any>;
  thumbnail_url?: string;
  images?: string[];
  average_rating: number;
  rating_count: number;
  is_featured: boolean;
  is_on_sale: boolean;
  is_active: boolean;
  discount_percentage: number;
  is_in_stock: boolean;
  is_low_stock: boolean;
  created_at: string;
  updated_at?: string;
}

export interface ProductListResponse {
  products: Product[];
  total: number;
  page: number;
  page_size: number;
  total_pages: number;
}

export interface ProductFilters {
  category?: ProductCategory;
  brand?: string;
  search?: string;
  is_featured?: boolean;
  is_on_sale?: boolean;
  min_price?: number;
  max_price?: number;
  in_stock_only?: boolean;
}

// ===== AI RECOMMENDATION INTERFACES =====

export interface RecommendationRequest {
  user_context?: Record<string, any>;
  category?: ProductCategory;
  limit?: number;
}

export interface RecommendedProduct {
  product: Product;
  confidence_score: number;
  recommendation_reason: string;
  brand_match: boolean;
  personalization_factors: string[];
}

export interface RecommendationResponse {
  recommendations: RecommendedProduct[];
  total: number;
  algorithm: string;
  personalized: boolean;
}

// ===== CART INTERFACES =====

export interface CartItemAdd {
  product_id: number;
  quantity?: number;
}

export interface CartItemUpdate {
  quantity: number;
}

export interface CartItemResponse {
  id: number;
  product_id: number;
  quantity: number;
  product: Product;
  subtotal: number;
  created_at: string;
}

export interface Cart {
  id: number;
  user_id: number;
  items: CartItemResponse[];
  total_items: number;
  subtotal: number;
  created_at: string;
  updated_at?: string;
}

// ===== ORDER INTERFACES =====

export interface OrderItemCreate {
  product_id: number;
  quantity: number;
}

export interface OrderItem {
  id: number;
  product_id: number;
  product_name: string;
  product_sku?: string;
  quantity: number;
  unit_price: number;
  subtotal: number;
  created_at: string;
}

export interface ShippingAddress {
  name: string;
  email: string;
  phone?: string;
  address_line1: string;
  address_line2?: string;
  city: string;
  state: string;
  zip: string;
  country: string;
}

export interface OrderCreate {
  items: OrderItemCreate[];
  shipping_address: ShippingAddress;
  billing_address?: ShippingAddress;
  customer_notes?: string;
}

export interface Order {
  id: number;
  order_number: string;
  user_id: number;
  status: OrderStatus;
  subtotal: number;
  tax: number;
  shipping_cost: number;
  discount: number;
  total: number;
  shipping_name: string;
  shipping_email: string;
  shipping_phone?: string;
  shipping_address_line1: string;
  shipping_address_line2?: string;
  shipping_city: string;
  shipping_state: string;
  shipping_zip: string;
  shipping_country: string;
  payment_method?: string;
  payment_status?: string;
  tracking_number?: string;
  tracking_url?: string;
  customer_notes?: string;
  created_at: string;
  updated_at?: string;
  shipped_at?: string;
  delivered_at?: string;
  items: OrderItem[];
}

export interface OrderListResponse {
  orders: Order[];
  total: number;
  page: number;
  page_size: number;
}

// ===== REVIEW INTERFACES =====

export interface ProductReviewCreate {
  product_id: number;
  rating: number;
  title?: string;
  comment?: string;
}

export interface ProductReviewUpdate {
  rating?: number;
  title?: string;
  comment?: string;
}

export interface ProductReview {
  id: number;
  product_id: number;
  user_id: number;
  rating: number;
  title?: string;
  comment?: string;
  is_verified_purchase: boolean;
  is_approved: boolean;
  helpful_count: number;
  created_at: string;
  updated_at?: string;
}

export interface ProductReviewListResponse {
  reviews: ProductReview[];
  total: number;
  average_rating: number;
}

// ===== HELPER FUNCTIONS =====

export const getCategoryIcon = (category: ProductCategory): string => {
  const iconMap: Record<ProductCategory, string> = {
    [ProductCategory.SUPPLEMENTS]: 'fitness',
    [ProductCategory.EQUIPMENT]: 'barbell',
    [ProductCategory.APPAREL]: 'shirt',
    [ProductCategory.ACCESSORIES]: 'watch',
    [ProductCategory.NUTRITION]: 'restaurant',
    [ProductCategory.RECOVERY]: 'medical',
    [ProductCategory.TECH]: 'phone-portrait',
    [ProductCategory.OTHER]: 'cube',
  };
  return iconMap[category] || 'cube';
};

export const getCategoryLabel = (category: ProductCategory): string => {
  const labelMap: Record<ProductCategory, string> = {
    [ProductCategory.SUPPLEMENTS]: 'Supplements',
    [ProductCategory.EQUIPMENT]: 'Equipment',
    [ProductCategory.APPAREL]: 'Apparel',
    [ProductCategory.ACCESSORIES]: 'Accessories',
    [ProductCategory.NUTRITION]: 'Nutrition',
    [ProductCategory.RECOVERY]: 'Recovery',
    [ProductCategory.TECH]: 'Tech & Gadgets',
    [ProductCategory.OTHER]: 'Other',
  };
  return labelMap[category] || 'Unknown';
};

export const getOrderStatusColor = (status: OrderStatus): string => {
  const colorMap: Record<OrderStatus, string> = {
    [OrderStatus.PENDING]: '#F59E0B',
    [OrderStatus.PROCESSING]: '#3B82F6',
    [OrderStatus.SHIPPED]: '#8B5CF6',
    [OrderStatus.DELIVERED]: '#10B981',
    [OrderStatus.CANCELLED]: '#EF4444',
    [OrderStatus.REFUNDED]: '#6B7280',
  };
  return colorMap[status] || '#6B7280';
};

export const getOrderStatusLabel = (status: OrderStatus): string => {
  const labelMap: Record<OrderStatus, string> = {
    [OrderStatus.PENDING]: 'Pending',
    [OrderStatus.PROCESSING]: 'Processing',
    [OrderStatus.SHIPPED]: 'Shipped',
    [OrderStatus.DELIVERED]: 'Delivered',
    [OrderStatus.CANCELLED]: 'Cancelled',
    [OrderStatus.REFUNDED]: 'Refunded',
  };
  return labelMap[status] || 'Unknown';
};

export const formatPrice = (price: number): string => {
  return `$${price.toFixed(2)}`;
};

export const calculateSavings = (price: number, comparePrice?: number): number => {
  if (!comparePrice || comparePrice <= price) return 0;
  return comparePrice - price;
};

export const getStarRating = (rating: number): string => {
  const fullStars = Math.floor(rating);
  const hasHalfStar = rating % 1 >= 0.5;
  let stars = '★'.repeat(fullStars);
  if (hasHalfStar) stars += '½';
  const emptyStars = 5 - Math.ceil(rating);
  stars += '☆'.repeat(emptyStars);
  return stars;
};

export const isPremiumBrand = (brand?: string): boolean => {
  if (!brand) return false;
  const premiumBrands = [
    'nike',
    'adidas',
    'under armour',
    'gymshark',
    'lululemon',
    'reebok',
    'puma',
    'new balance',
    'asics',
    'mizuno',
    'brooks',
    'hoka',
    'on running',
  ];
  return premiumBrands.includes(brand.toLowerCase());
};

File: backend/app/models/shop.py
"""
Shop Models - Product catalog, orders, and recommendations
"""
from sqlalchemy import Column, Integer, String, Float, Boolean, ForeignKey, DateTime, Text, JSON, Enum as SQLEnum
from sqlalchemy.orm import relationship
from datetime import datetime
import enum

from app.core.database import Base


class ProductCategory(enum.Enum):
    """Product categories"""
    SUPPLEMENTS = "supplements"
    EQUIPMENT = "equipment"
    APPAREL = "apparel"
    ACCESSORIES = "accessories"
    NUTRITION = "nutrition"
    RECOVERY = "recovery"
    TECH = "tech"
    OTHER = "other"


class OrderStatus(enum.Enum):
    """Order status"""
    PENDING = "pending"
    PROCESSING = "processing"
    SHIPPED = "shipped"
    DELIVERED = "delivered"
    CANCELLED = "cancelled"
    REFUNDED = "refunded"


class Product(Base):
    """Product model"""
    __tablename__ = "products"

    id = Column(Integer, primary_key=True, index=True, autoincrement=True)
    name = Column(String(255), nullable=False, index=True)
    slug = Column(String(300), unique=True, nullable=False, index=True)
    description = Column(Text, nullable=True)
    short_description = Column(String(500), nullable=True)

    category = Column(SQLEnum(ProductCategory), nullable=False, index=True)
    brand = Column(String(100), nullable=True, index=True)

    price = Column(Float, nullable=False)
    compare_at_price = Column(Float, nullable=True)  # Original price for sales
    cost = Column(Float, nullable=True)  # Cost to store

    sku = Column(String(100), unique=True, nullable=True)
    barcode = Column(String(100), nullable=True)

    stock_quantity = Column(Integer, default=0)
    low_stock_threshold = Column(Integer, default=10)

    weight_kg = Column(Float, nullable=True)
    dimensions = Column(JSON, nullable=True)  # {length, width, height}

    images = Column(JSON, nullable=True)  # Array of image URLs
    thumbnail_url = Column(String(500), nullable=True)

    # SEO
    meta_title = Column(String(255), nullable=True)
    meta_description = Column(String(500), nullable=True)

    # Features & specs
    features = Column(JSON, nullable=True)  # Array of feature strings
    specifications = Column(JSON, nullable=True)  # Dict of specs

    # Ratings
    average_rating = Column(Float, default=0.0)
    rating_count = Column(Integer, default=0)

    # Status
    is_active = Column(Boolean, default=True)
    is_featured = Column(Boolean, default=False)
    is_on_sale = Column(Boolean, default=False)

    # Timestamps
    created_at = Column(DateTime, default=datetime.utcnow)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)

    # Relationships
    order_items = relationship("OrderItem", back_populates="product")
    reviews = relationship("ProductReview", back_populates="product", cascade="all, delete-orphan")

    @property
    def discount_percentage(self):
        """Calculate discount percentage"""
        if self.compare_at_price and self.compare_at_price > self.price:
            return round(((self.compare_at_price - self.price) / self.compare_at_price) * 100)
        return 0

    @property
    def is_in_stock(self):
        """Check if product is in stock"""
        return self.stock_quantity > 0

    @property
    def is_low_stock(self):
        """Check if product is low on stock"""
        return 0 < self.stock_quantity <= self.low_stock_threshold


class Order(Base):
    """Order model"""
    __tablename__ = "orders"

    id = Column(Integer, primary_key=True, index=True, autoincrement=True)
    order_number = Column(String(50), unique=True, nullable=False, index=True)

    user_id = Column(Integer, ForeignKey("users.id"), nullable=False, index=True)

    # Order details
    status = Column(SQLEnum(OrderStatus), default=OrderStatus.PENDING, index=True)
    subtotal = Column(Float, nullable=False)
    tax = Column(Float, default=0.0)
    shipping_cost = Column(Float, default=0.0)
    discount = Column(Float, default=0.0)
    total = Column(Float, nullable=False)

    # Shipping info
    shipping_name = Column(String(255), nullable=False)
    shipping_email = Column(String(255), nullable=False)
    shipping_phone = Column(String(50), nullable=True)
    shipping_address_line1 = Column(String(255), nullable=False)
    shipping_address_line2 = Column(String(255), nullable=True)
    shipping_city = Column(String(100), nullable=False)
    shipping_state = Column(String(100), nullable=False)
    shipping_zip = Column(String(20), nullable=False)
    shipping_country = Column(String(100), nullable=False)

    # Billing info (can be same as shipping)
    billing_address_line1 = Column(String(255), nullable=True)
    billing_address_line2 = Column(String(255), nullable=True)
    billing_city = Column(String(100), nullable=True)
    billing_state = Column(String(100), nullable=True)
    billing_zip = Column(String(20), nullable=True)
    billing_country = Column(String(100), nullable=True)

    # Payment
    payment_method = Column(String(50), nullable=True)
    payment_status = Column(String(50), nullable=True)
    payment_transaction_id = Column(String(255), nullable=True)

    # Notes
    customer_notes = Column(Text, nullable=True)
    internal_notes = Column(Text, nullable=True)

    # Tracking
    tracking_number = Column(String(255), nullable=True)
    tracking_url = Column(String(500), nullable=True)

    # Timestamps
    created_at = Column(DateTime, default=datetime.utcnow)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)
    shipped_at = Column(DateTime, nullable=True)
    delivered_at = Column(DateTime, nullable=True)

    # Relationships
    user = relationship("User", back_populates="orders")
    items = relationship("OrderItem", back_populates="order", cascade="all, delete-orphan")


class OrderItem(Base):
    """Order item model"""
    __tablename__ = "order_items"

    id = Column(Integer, primary_key=True, index=True, autoincrement=True)
    order_id = Column(Integer, ForeignKey("orders.id"), nullable=False, index=True)
    product_id = Column(Integer, ForeignKey("products.id"), nullable=False, index=True)

    product_name = Column(String(255), nullable=False)  # Snapshot at time of order
    product_sku = Column(String(100), nullable=True)

    quantity = Column(Integer, nullable=False)
    unit_price = Column(Float, nullable=False)
    subtotal = Column(Float, nullable=False)

    # Timestamps
    created_at = Column(DateTime, default=datetime.utcnow)

    # Relationships
    order = relationship("Order", back_populates="items")
    product = relationship("Product", back_populates="order_items")


class ProductReview(Base):
    """Product review model"""
    __tablename__ = "product_reviews"

    id = Column(Integer, primary_key=True, index=True, autoincrement=True)
    product_id = Column(Integer, ForeignKey("products.id"), nullable=False, index=True)
    user_id = Column(Integer, ForeignKey("users.id"), nullable=False, index=True)

    rating = Column(Integer, nullable=False)  # 1-5
    title = Column(String(255), nullable=True)
    comment = Column(Text, nullable=True)

    is_verified_purchase = Column(Boolean, default=False)
    is_approved = Column(Boolean, default=True)

    # Helpfulness
    helpful_count = Column(Integer, default=0)

    # Timestamps
    created_at = Column(DateTime, default=datetime.utcnow)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)

    # Relationships
    product = relationship("Product", back_populates="reviews")
    user = relationship("User", back_populates="product_reviews")


class ShoppingCart(Base):
    """Shopping cart model"""
    __tablename__ = "shopping_carts"

    id = Column(Integer, primary_key=True, index=True, autoincrement=True)
    user_id = Column(Integer, ForeignKey("users.id"), nullable=False, unique=True, index=True)

    # Timestamps
    created_at = Column(DateTime, default=datetime.utcnow)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)

    # Relationships
    user = relationship("User", back_populates="shopping_cart")
    items = relationship("CartItem", back_populates="cart", cascade="all, delete-orphan")


class CartItem(Base):
    """Cart item model"""
    __tablename__ = "cart_items"

    id = Column(Integer, primary_key=True, index=True, autoincrement=True)
    cart_id = Column(Integer, ForeignKey("shopping_carts.id"), nullable=False, index=True)
    product_id = Column(Integer, ForeignKey("products.id"), nullable=False, index=True)

    quantity = Column(Integer, nullable=False, default=1)

    # Timestamps
    created_at = Column(DateTime, default=datetime.utcnow)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)

    # Relationships
    cart = relationship("ShoppingCart", back_populates="items")
    product = relationship("Product")
